include/revng/Pipeline/Context.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <initializer_list>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Casting.h"
+
+#include "revng/Pipeline/KindsRegistry.h"
+#include "revng/Pipeline/SavableObject.h"
+
+namespace Pipeline {
+
+/// a simple pair that holds a savable object a name meant to describe a object
+/// which lifetime must be than the context.
+class NamedGlobalReference {
+private:
+  llvm::StringRef Name;
+  SavableObjectBase *Obj;
+
+public:
+  NamedGlobalReference(llvm::StringRef Name, SavableObjectBase &Obj) :
+    Name(Name), Obj(&Obj) {}
+
+  llvm::StringRef name() const { return Name; }
+
+  SavableObjectBase &value() const { return *Obj; }
+};
+
+/// A class that contains every object that has a lifetime longer
+/// than a pipeline.
+///
+/// This includes the kinds and the NamedGlobals that will be available in that
+/// context
+class Context {
+private:
+  llvm::StringMap<SavableObjectBase *> Globals;
+  KindsRegistry TheKindRegistry;
+
+public:
+  explicit Context(KindsRegistry registry) :
+    TheKindRegistry(std::move(registry)) {}
+  Context();
+  Context(llvm::ArrayRef<NamedGlobalReference> Globals, KindsRegistry registry);
+  Context(llvm::ArrayRef<NamedGlobalReference> Globals);
+  Context(const Context &) = default;
+  Context(Context &&) = default;
+  Context &operator=(const Context &) = default;
+  Context &operator=(Context &&) = default;
+  ~Context() = default;
+
+  template<typename T>
+  llvm::Expected<T *> getGlobal(llvm::StringRef Name) const {
+    auto Iter = Globals.find(Name);
+    if (Iter == Globals.end())
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "pipeline loader context did not "
+                                     "contained object %s",
+                                     Name.str().c_str());
+    auto *Casted = llvm::dyn_cast<T>(Iter->second);
+    if (Casted == nullptr)
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "pipeline loader was requested to cast %s "
+                                     "to the wrong type",
+                                     Name.str().c_str());
+
+    return Casted;
+  }
+
+  void addGlobal(llvm::StringRef Name, SavableObjectBase &Object) {
+    Globals.try_emplace(Name, &Object);
+  }
```

Comment:
Unused. Let's strive for a minimal API.
```suggestion
  void addGlobal(llvm::StringRef Name, SavableObjectBase &Object) {
    Globals.try_emplace(Name, &Object);
  }
```